### PR TITLE
[WIP] : Multiple project owner backend.

### LINKED ISSUE #3958: Support multiple project owners
In order to better manage Litmus, we would need to extend the current behaviour by allowing multiple projects owners. This is especially useful when integrating Litmus within SSO (single sign-on via OIDC) and where we need someone within the organisation to became the project owner in addition to the administrator created by default.

The purposed feature implementation can be described as follow.

Any created users can be invited into a project with the owner role. Then the default administrator will remain but not being used (not mapped to any user in the organisation).

Feedback appreciated.

A PR is attached to reflect this feature proposal.

### PR DESCRIPTION
**Pull request by aryan-bhokare**
closes #3958 #4161 

## Proposed changes

Work done: 
1. Modification of db schema.
2.  Addition of new API `GetProjectOwners`.
3. Modifications of `SendInvitation` API.
4. Modifications of `LeaveProject` API.
5. Added new API `UpdateMemberRole`.
6. Added new API `DeleteProject`.

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [x] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the commit for DCO to be passed.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added necessary documentation (if appropriate)

## Dependency
- Please add the links to the dependent PR need to be merged before this (if any).

### CODE REVIEW

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 1]
```diff
@@ -195,6 +195,30 @@ func GetActiveProjectMembers(service services.ApplicationService) gin.HandlerFun
...
+//	@Param			state	path	string	true	"State"
+//	@Accept			json
+//	@Produce		json
+//	@Failure		500	{object}	response.ErrServerError
+//	@Success		200	{object}	response.Response{}
+//	@Router			/get_project_owners/:project_id/:state [get]
+func GetActiveProjectOwners(service services.ApplicationService) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		projectID := c.Param("project_id")
+		// state := c.Param("state")
```
Comment (Saranya-jena) [inline chaoscenter/authentication/api/handlers/rest/project_handler.go:212]: can remove if not required

File: chaoscenter/authentication/pkg/entities/project.go [hunk 2]
```diff
@@ -10,9 +10,13 @@ type Project struct {
 }
 
 type Owner struct {
-	UserID   string `bson:"user_id" json:"userID"`
-	Username string `bson:"username" json:"username"`
+	UserID        string     `bson:"user_id" json:"userID"`
```
Comment (Saranya-jena) [inline chaoscenter/authentication/pkg/entities/project.go:13]: make it memory efficient. Ref: https://towardsdev.com/golang-writing-memory-efficient-and-cpu-optimized-go-structs-62fcef4dbfd0

File: chaoscenter/web/yarn.lock [hunk 3]
```diff
@@ -4325,9 +4325,9 @@ flatted@^3.2.9:
   integrity sha512-X8cqMLLie7KsNUDSdzeN8FYK9rEt4Dt67OsG/DNGnYTSDBG4uFAJFBnUeiV+zCVAvwFy56IjM9sH51jVaEhNxw==
 
 follow-redirects@^1.0.0, follow-redirects@^1.15.0:
-  version "1.15.5"
-  resolved "https://registry.yarnpkg.com/follow-redirects/-/follow-redirects-1.15.5.tgz#54d4d6d062c0fa7d9d17feb008461550e3ba8020"
-  integrity sha512-vSFWUON1B+yAw1VN4xMfxgn5fTUiaOzAJCKBwIIgT/+7CuGy9+r+5gITvP62j3RmaD5Ph65UaERdOSRGUzZtgw==
+  version "1.15.6"
```
Comment (Saranya-jena) [inline chaoscenter/web/yarn.lock:4328]: why is this change required?

**Review by Saranya-jena — CHANGES_REQUESTED** [review]

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 1 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:212]: Sure

File: chaoscenter/web/yarn.lock (hunk 3 above)
Comment (aryan-bhokare) [reply chaoscenter/web/yarn.lock:4328]: It happened automatically. I am unable to understand I was thinking about starting new from the commit before the bumps

Comment (aryan-bhokare) [reply chaoscenter/web/yarn.lock:4328]: these changes were included in the main litmus repo I will remove it.

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 4]
```diff
@@ -569,6 +593,20 @@ func LeaveProject(service services.ApplicationService) gin.HandlerFunc {
 			return
 		}
 
+		if member.Role == nil || *member.Role == entities.RoleOwner {
```
Comment (Saranya-jena) [inline chaoscenter/authentication/api/handlers/rest/project_handler.go:596]: Why are we checking `member.Role == nil`

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 5]
```diff
@@ -726,6 +764,55 @@ func UpdateProjectName(service services.ApplicationService) gin.HandlerFunc {
...
+			service)
+		if err != nil {
+			log.Warn(err)
+			c.JSON(utils.ErrorStatusCodes[utils.ErrUnauthorized],
+				presenter.CreateErrorResponse(utils.ErrUnauthorized))
+			return
+		}
+
+		err = service.UpdateMemberRole(member.ProjectID, member.UserID, member.Role)
+		if err != nil {
```
Comment (Saranya-jena) [inline chaoscenter/authentication/api/handlers/rest/project_handler.go:804]: I think we should also add one more check for a case: There is only 1 owner and that owner is trying update their role to viewer/editor. In this case there will be no project owner. 🤔

File: chaoscenter/authentication/pkg/project/repository.go [hunk 6]
```diff
@@ -277,6 +279,24 @@
...
+func (r repository) UpdateMemberRole(projectID string, userID string, role *entities.MemberRole) error {
+	opts := options.Update().SetArrayFilters(options.ArrayFilters{
+		Filters: []interface{}{
+			bson.D{{"elem.user_id", userID}},
+		},
+	})
+	query := bson.D{{"_id", projectID}}
+	update := bson.D{{"$set", bson.M{"members.$[elem]role": role}}}
+
+	_, err := r.Collection.UpdateOne(context.TODO(), query, update, opts)
```
[the remark this answers is not in the document]
Comment (Saranya-jena) [reply chaoscenter/authentication/pkg/project/repository.go:292]: same

**Review by Saranya-jena — CHANGES_REQUESTED** [review]

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 5 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:804]: sure I'll do that

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 4 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:595]: because if we don't, it gives a runtime error attempting to access or deference nil pointer

Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:595]:
```
2024/04/09 14:28:53 [Recovery] 2024/04/09 - 14:28:53 panic recovered:
runtime error: invalid memory address or nil pointer dereference
/usr/local/go/src/runtime/panic.go:261 (0x4556d7)
	panicmem: panic(memoryError)
/usr/local/go/src/runtime/signal_unix.go:881 (0x4556a5)
	sigpanic: panicmem()
/home/aryan/Desktop/dev2/lfx/litmus/chaoscenter/authentication/api/handlers/rest/project_handler.go:595 (0xc540df)
	ProjectRouter.LeaveProject.func16: if *member.Role == entities.RoleOwner {
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xc1972a)
	(*Context).Next: c.handlers[c.index](c)
/home/aryan/Desktop/dev2/lfx/litmus/chaoscenter/authentication/api/middleware/jwt_middlware.go:33 (0xc5824e)
	ProjectRouter.JwtMiddleware.func1: c.Next()
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xc1972a)
	(*Context).Next: c.handlers[c.index](c)
/home/aryan/Desktop/dev2/lfx/litmus/chaoscenter/authentication/api/middleware/jwt_middlware.go:33 (0xc5b54e)
	UserRouter.JwtMiddleware.func3: c.Next()
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xc26619)
	(*Context).Next: c.handlers[c.index](c)
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/recovery.go:102 (0xc26607)
	CustomRecoveryWithWriter.func1: c.Next()
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xc2575c)
	(*Context).Next: c.handlers[c.index](c)
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/logger.go:240 (0xc25743)
	LoggerWithConfig.func1: c.Next()
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174 (0xc24c4d)
	(*Context).Next: c.handlers[c.index](c)
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:620 (0xc248dc)
	(*Engine).handleHTTPRequest: c.Next()
/home/aryan/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:576 (0xc24411)
	(*Engine).ServeHTTP: engine.handleHTTPRequest(c)
/usr/local/go/src/net/http/server.go:3137 (0x800ced)
	serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
/usr/local/go/src/net/http/server.go:2039 (0x7fbfc7)
	(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
/usr/local/go/src/runtime/asm_amd64.s:1695 (0x472c20)
	goexit: BYTE	$0x90	// NOP


```

File: chaoscenter/authentication/pkg/entities/project.go (hunk 2 above)
Comment (SarthakJain26) [reply chaoscenter/authentication/pkg/entities/project.go:13]: Username is an mandatory field, making it pointer does not makes sense

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 5 above)
Comment (SarthakJain26) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:803]: How about a check to not allow the logged in user to change their own project roles. This will cover all the conditions. Please check on feasibility of this approach and race conditions if any

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 7]
```diff
@@ -569,6 +592,20 @@ func LeaveProject(service services.ApplicationService) gin.HandlerFunc {
 			return
 		}
 
+		if member.Role == nil || *member.Role == entities.RoleOwner {
```
Comment (SarthakJain26) [inline chaoscenter/authentication/api/handlers/rest/project_handler.go:595]: The check should be instead
```suggestion
		if member.Role != nil && *member.Role == entities.RoleOwner {
```

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 8]
```diff
@@ -796,3 +882,50 @@ func GetProjectRole(service services.ApplicationService) gin.HandlerFunc {
...
+//	@Failure		400	{object}	response.ErrProjectNotFound
+//	@Failure		500	{object}	response.ErrServerError
+//	@Success		200	{object}	response.Response{}
+//	@Router			/delete_project [post]
+//
+// DeleteProject is used to delete a project.
+func DeleteProject (service services.ApplicationService) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		var deleteProjectRequest entities.DeleteProjectInput
+		err := c.BindJSON(&deleteProjectRequest)
```
Comment (SarthakJain26) [inline chaoscenter/authentication/api/handlers/rest/project_handler.go:901]: If delete_project takes just one parameter i.e project_id, better to take it from the URL itself instead of taking it from the request body. `/delete_project/{project_id}`

**Review by SarthakJain26 — CHANGES_REQUESTED** [review]

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 8 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:901]: on it.

File: chaoscenter/authentication/api/handlers/rest/project_handler.go [hunk 9]
```diff
@@ -796,3 +882,50 @@ func GetProjectRole(service services.ApplicationService) gin.HandlerFunc {
...
+//	@Tags			ProjectRouter
+//	@Accept			json
+//	@Produce		json
+//	@Failure		400	{object}	response.ErrProjectNotFound
+//	@Failure		500	{object}	response.ErrServerError
+//	@Success		200	{object}	response.Response{}
+//	@Router			/delete_project [post]
+//
+// DeleteProject is used to delete a project.
+func DeleteProject (service services.ApplicationService) gin.HandlerFunc {
```
[the remark this answers is not in the document]
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go:898]: Lets do that in different pr.

File: chaoscenter/authentication/pkg/entities/project.go (hunk 2 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/pkg/entities/project.go:13]: hey @SarthakJain26 better align give many more suggestions to save memory should I modify them too or stick with the schema I wrote.

File: chaoscenter/authentication/api/handlers/rest/project_handler.go (hunk 7 above)
Comment (aryan-bhokare) [reply chaoscenter/authentication/api/handlers/rest/project_handler.go]: done

**Review by SarthakJain26 — APPROVED** [review]
Looks good to me 🚀 , great work

**Comment by aryan-bhokare** [comment]
> Looks good to me 🚀 , great work

Thank you 🚀
